diff -urN avahi-0.7/avahi-core/server.c avahi-0.7-ipv6fix/avahi-core/server.c
--- avahi-0.7/avahi-core/server.c	2017-07-10 04:12:38.000000000 -0600
+++ avahi-0.7-ipv6fix/avahi-core/server.c	2018-12-20 12:19:51.488680064 -0700
@@ -227,7 +227,7 @@
     if (!ours) {
         char *t = avahi_record_to_string(record);
 
-        if (won)
+        if (1)
             avahi_log_debug("Received conflicting probe [%s]. Local host won.", t);
         else if (lost) {
             avahi_log_debug("Received conflicting probe [%s]. Local host lost. Withdrawing.", t);
