diff -urN a/tmpfiles.sh b/tmpfiles.sh
--- a/tmpfiles.sh	2020-11-21 17:17:21.869243322 -0300
+++ b/tmpfiles.sh	2020-11-21 17:18:13.205241217 -0300
@@ -59,7 +59,7 @@
 }
 
 _setfacl() {
-	dryrun_or_real setfacl -P "$1" "$2" "$3" -- "$4"
+	dryrun_or_real setfacl "$1" "$2" "$3" -- "$4"
 }
 
 relabel() {
@@ -293,7 +293,7 @@
 	# The format of the argument field matches setfacl
 	local ACTION='--remove-all --set'
 	[ "$FORCE" -gt 0 ] && ACTION='--modify'
-	_setfacl '' "$ACTION" "$6" "$1"
+	_setfacl '-P' "$ACTION" "$6" "$1"
 }
 
 _A() {
@@ -302,7 +302,7 @@
 	# Does not follow symlinks
 	local ACTION='--remove-all --set'
 	[ "$FORCE" -gt 0 ] && ACTION='--modify'
-	_setfacl -R "$ACTION" "$6" "$1"
+	_setfacl '-P -R' "$ACTION" "$6" "$1"
 }
 
 _h() {
