--- linux-4.19.66-gentoo/drivers/gpu/drm/amd/amdgpu/amdgpu_drv.c	2019-08-26 06:23:22.455074755 +0200
+++ linux-4.19.66-gentoo/drivers/gpu/drm/amd/amdgpu/amdgpu_drv.c	2019-09-09 13:11:52.879809819 +0200
@@ -785,6 +785,7 @@ static const struct pci_device_id pciidl
 	{0x1002, 0x66AF, PCI_ANY_ID, PCI_ANY_ID, 0, 0, CHIP_VEGA20|AMD_EXP_HW_SUPPORT},
 	/* Raven */
 	{0x1002, 0x15dd, PCI_ANY_ID, PCI_ANY_ID, 0, 0, CHIP_RAVEN|AMD_IS_APU},
+	{0x1002, 0x15d8, PCI_ANY_ID, PCI_ANY_ID, 0, 0, CHIP_RAVEN|AMD_IS_MOBILITY|AMD_IS_APU},
 
 	{0, 0, 0}
 };
--- linux-4.19.66-gentoo/drivers/gpu/drm/amd/amdkfd/kfd_device.c	2019-08-26 06:23:22.683079348 +0200
+++ linux-4.19.66-gentoo/drivers/gpu/drm/amd/amdkfd/kfd_device.c	2019-09-09 13:31:18.563024146 +0200
@@ -275,6 +275,7 @@ static const struct kfd_deviceid support
 	{ 0x9876, &carrizo_device_info },	/* Carrizo */
 	{ 0x9877, &carrizo_device_info },	/* Carrizo */
 	{ 0x15DD, &raven_device_info },		/* Raven */
+	{ 0x15D8, &raven_device_info },		/* Raven */
 #endif
 	{ 0x67A0, &hawaii_device_info },	/* Hawaii */
 	{ 0x67A1, &hawaii_device_info },	/* Hawaii */
