diff -urN a/cnf/sets/portage.conf b/cnf/sets/portage.conf
--- a/cnf/sets/portage.conf	2020-11-24 14:20:26.303049372 -0300
+++ b/cnf/sets/portage.conf	2020-11-24 16:48:26.266685223 -0300
@@ -71,6 +71,7 @@
 [module-rebuild]
 class = portage.sets.dbapi.OwnerSet
 files = /lib/modules
+exclude-files = /usr/src
 
 # Installed packages that own files inside /usr/lib/xorg/modules,
 # excluding the package that owns /usr/bin/Xorg.
