diff -urN a/lib/portage/_sets/dbapi.py b/lib/portage/_sets/dbapi.py
--- a/lib/portage/_sets/dbapi.py	2020-11-24 14:20:26.315049372 -0300
+++ b/lib/portage/_sets/dbapi.py	2021-04-09 08:17:37.137892499 -0300
@@ -85,6 +85,8 @@
 		if exclude_paths is None:
 			for link, p in vardb._owners.iter_owners(paths):
 				pkg = pkg_str(link.mycpv, None)
+				if "sys-kernel/debian" in pkg.cp:
+					continue
 				rValue.add("%s:%s" % (pkg.cp, pkg.slot))
 		else:
 			all_paths = set()
@@ -94,6 +96,8 @@
 			for link, p in vardb._owners.iter_owners(all_paths):
 				pkg = pkg_str(link.mycpv, None)
 				atom = "%s:%s" % (pkg.cp, pkg.slot)
+				if "sys-kernel/debian" in pkg.cp:
+					continue
 				rValue.add(atom)
 				# Returned paths are relative to ROOT and do not have
 				# a leading slash.
