diff -urN portage-3.0.9/bin/ebuild.sh portage-3.0.9-fixed/bin/ebuild.sh
--- portage-3.0.9/bin/ebuild.sh	2021-02-19 20:39:11.918915396 -0700
+++ portage-3.0.9-fixed/bin/ebuild.sh	2021-02-19 20:41:45.938911106 -0700
@@ -677,10 +677,10 @@
 		# Make it group writable. 666&~002==664
 		umask 002
 	fi
-
+	_PY_COMPAT="${PYTHON_COMPAT[@]}"
 	auxdbkeys="DEPEND RDEPEND SLOT SRC_URI RESTRICT HOMEPAGE LICENSE
 		DESCRIPTION KEYWORDS INHERITED IUSE REQUIRED_USE PDEPEND BDEPEND
-		EAPI PROPERTIES DEFINED_PHASES UNUSED_05 UNUSED_04
+		EAPI PROPERTIES DEFINED_PHASES UNUSED_05 _PY_COMPAT
 		UNUSED_03 UNUSED_02 UNUSED_01"
 
 	if ! ___eapi_has_BDEPEND; then
@@ -699,6 +699,7 @@
 		done
 		eval "exec ${PORTAGE_PIPE_FD}>&-"
 	fi
+	unset _PY_COMPAT
 	set +f
 else
 	# Note: readonly variables interfere with __preprocess_ebuild_env(), so
