diff -urN portage-2.3.68/cnf/make.globals portage-2.3.68-fixed/cnf/make.globals
--- portage-2.3.68/cnf/make.globals	2019-07-08 22:38:58.177027829 -0600
+++ portage-2.3.68-fixed/cnf/make.globals	2019-07-08 22:41:33.090556025 -0600
@@ -27,10 +27,10 @@
 ACCEPT_RESTRICT="*"
 
 # Miscellaneous paths
-DISTDIR="/var/cache/distfiles"
-PKGDIR="/var/cache/binpkgs"
-RPMDIR="/var/cache/rpm"
-
+DISTDIR="/var/cache/portage/distfiles"
+PKGDIR="/var/cache/portage/packages"
+RPMDIR="/var/cache/portage/rpm"
+ 
 # Temporary build directory
 PORTAGE_TMPDIR="/var/tmp"
 
