diff -urN a/Modules/FindLAPACK.cmake b/Modules/FindLAPACK.cmake
--- a/Modules/FindLAPACK.cmake	2020-11-18 22:04:23.845347306 -0300
+++ b/Modules/FindLAPACK.cmake	2020-11-18 22:13:03.141351641 -0300
@@ -7,6 +7,10 @@
 
 Find Linear Algebra PACKage (LAPACK) library
 
+Version modified for Gentoo Linux.
+If a valid PkgConfig configuration is found, this overrides and cancels
+all further checks.
+
 This module finds an installed Fortran library that implements the
 LAPACK linear-algebra interface (see http://www.netlib.org/lapack/).
 
@@ -224,6 +228,22 @@
 
 _lapack_find_library_setup()
 
+# first, try PkgConfig
+find_package(PkgConfig REQUIRED)
+pkg_check_modules(PC_LAPACK lapack)
+if(PC_LAPACK_FOUND)
+  foreach(PC_LIB ${PC_LAPACK_LIBRARIES})
+    find_library(${PC_LIB}_LIBRARY NAMES ${PC_LIB} HINTS ${PC_LAPACK_LIBRARY_DIRS} )
+    if (NOT ${PC_LIB}_LIBRARY)
+      message(FATAL_ERROR "Something is wrong in your pkg-config file - lib ${PC_LIB} not found in ${PC_LAPACK_LIBRARY_DIRS}")
+    endif (NOT ${PC_LIB}_LIBRARY)
+    list(APPEND LAPACK_LIBRARIES ${${PC_LIB}_LIBRARY})
+  endforeach(PC_LIB)
+  find_package_handle_standard_args(LAPACK DEFAULT_MSG LAPACK_LIBRARIES)
+  mark_as_advanced(LAPACK_LIBRARIES)
+else(PC_LAPACK_FOUND)
+message(STATUS "No PkgConfig configuration for LAPACK found; starting more extensive search.")
+
 set(LAPACK_LINKER_FLAGS)
 set(LAPACK_LIBRARIES)
 set(LAPACK95_LIBRARIES)
@@ -553,3 +573,4 @@
 endif()
 
 _lapack_find_library_teardown()
+endif(PC_LAPACK_FOUND)
