From 69298ebfc2c066acee5d187eac8ce9f38c796630 Mon Sep 17 00:00:00 2001
From: Darren Tucker <dtucker@dtucker.net>
Date: Wed, 13 Nov 2019 23:27:31 +1100
Subject: Remove duplicate __NR_clock_nanosleep

---
 sandbox-seccomp-filter.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/sandbox-seccomp-filter.c b/sandbox-seccomp-filter.c
index 96ab141f..be239767 100644
--- a/sandbox-seccomp-filter.c
+++ b/sandbox-seccomp-filter.c
@@ -245,9 +245,6 @@ static const struct sock_filter preauth_insns[] = {
 #ifdef __NR_clock_nanosleep
 	SC_ALLOW(__NR_clock_nanosleep),
 #endif
-#ifdef __NR_clock_nanosleep
-	SC_ALLOW(__NR_clock_nanosleep),
-#endif
 #ifdef __NR__newselect
 	SC_ALLOW(__NR__newselect),
 #endif
-- 
cgit v1.2.3

