--- ruby.eselect-20190121	2019-05-13 16:22:41.211413287 -0600
+++ ruby.eselect-20190121	2019-05-13 16:25:07.083416822 -0600
@@ -44,6 +44,16 @@
 	rm -f ${EROOT}${bindir}/{ruby,gem,irb,erb,ri,rdoc} && \
 	rm -f ${EROOT}${man1dir}/{ruby,irb,erb,ri}.1*
 
+	# funtoo mod start
+	for x in ${EROOT}${bindir}/ruby[0-9].[0-9]; do
+		# remove ruby1.9 -> ruby19 symlink if it exists:
+		if [ -L "$x" ]; then
+			rm -f "$x"
+		fi
+	done
+	# funtoo mod end
+	
+
 	# Remove legacy testrb link if it still exists.
 	local testrb="${EROOT}${bindir}/testrb"
 	if [[ -L ${testrb} ]]; then
@@ -83,6 +93,15 @@
 			die -q "Could not set ${f} symlink"
 	done
 
+	# funtoo mod start
+	case $version in
+	[0-9][0-9])
+		# create ruby1.9 symlink if we have ruby19, for example:
+		ln -s "ruby${version}" "${path}/ruby${version:0:1}.${version:1:2}"
+		;;
+	esac
+	# funtoo mod end
+
 	# these can fail
 	if [[ -e "${path}gem${version}" ]] ; then
 		ln -s "gem${version}" "${path}gem" || \
