diff -urN linux-4.19.132/drivers/net/wireless/intel/iwlwifi/iwl-drv.c linux-4.19.132-fixed/drivers/net/wireless/intel/iwlwifi/iwl-drv.c
--- linux-4.19.132/drivers/net/wireless/intel/iwlwifi/iwl-drv.c	2020-08-15 21:27:26.960005058 -0600
+++ linux-4.19.132-fixed/drivers/net/wireless/intel/iwlwifi/iwl-drv.c	2020-08-15 21:34:37.864022784 -0600
@@ -1649,6 +1649,7 @@
 	.d0i3_disable = true,
 	.d0i3_timeout = 1000,
 	.uapsd_disable = IWL_DISABLE_UAPSD_BSS | IWL_DISABLE_UAPSD_P2P_CLIENT,
+	.disable_11n = IWL_DISABLE_HT_RXAGG,
 	/* the rest are 0 by default */
 };
 IWL_EXPORT_SYMBOL(iwlwifi_mod_params);
