diff -urN go/src/github.com/lxc/lxd/lxd/devices.go go-fixed/src/github.com/lxc/lxd/lxd/devices.go
--- go/src/github.com/lxc/lxd/lxd/devices.go	2020-02-13 15:08:48.000000000 -0700
+++ go-fixed/src/github.com/lxc/lxd/lxd/devices.go	2020-02-17 20:05:08.282321085 -0700
@@ -238,6 +238,7 @@
 				}
 
 				vendor, product, ok := ueventParseVendorProduct(props, subsystem, devname)
+				logger.Errorf("Parsed vendor/product for %s, got: %s/%s", devname, vendor, product)
 				if !ok {
 					continue
 				}
@@ -253,7 +254,12 @@
 				}
 
 				zeroPad := func(s string, l int) string {
+					if (len(s) <= 4) {
 					return strings.Repeat("0", l-len(s)) + s
+				} else {
+					logger.Errorf("Can't zeropad vendor/product: '%s'", s)
+					return s
+				}
 				}
 
 				unix, err := device.UnixHotplugNewEvent(
