diff -ru tcsh-6.24.01.orig/configure.ac tcsh-6.24.01/configure.ac
--- tcsh-6.24.01.orig/configure.ac	2022-08-30 08:46:40.716024572 +0900
+++ tcsh-6.24.01/configure.ac	2022-08-30 11:07:30.424433514 +0900
@@ -7,7 +7,7 @@
 dnl
 dnl Written by Kaveh Ghazi (ghazi@caip.rutgers.edu) 5/11/96.
 
-AC_PREREQ([2.71])dnl Minimum Autoconf version required.
+AC_PREREQ([2.69])dnl Minimum Autoconf version required.
 dnl UPDATE NEXT LINE FOR RELEASE
 AC_INIT([tcsh],[6.24.01],[https://bugs.astron.com/])
 AC_CONFIG_MACRO_DIR([m4])
